airflow/models/xcom_arg.py

```diff
@@ -0,0 +1,116 @@
+
+from airflow.configuration import conf
+from airflow.models.baseoperator import BaseOperator
+from airflow.models.xcom import XCOM_RETURN_KEY
+from airflow.utils.module_loading import import_string
+
+
+class BaseXComArg:
+    """
+    Class that represents a XCom push from a previous operator.
+    Defaults to "return_value" as only key.
+    """
+
+    def __init__(self, operator: BaseOperator, keys: Optional[List[str]] = None):
+        self._operator = operator
+        self._keys = keys or [XCOM_RETURN_KEY]
+
+    def __eq__(self, other):
+        return self._operator == other._operator  # pylint: disable=protected-access
+
+    def __lshift__(self, other):
+        self.set_upstream(other)
+        return self
+
+    def __rshift__(self, other):
+        self.set_downstream(other)
+        return self
+
+    def set_upstream(self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]):
+        """
+        Proxy to underlying operator set_upstream method
+        """
+        self._operator.set_upstream(task_or_task_list)
+
+    def set_downstream(
+        self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]
+    ):
+        """
+        Proxy to underlying operator set_downstream method
+        """
+        self._operator.set_downstream(task_or_task_list)
+
+    def resolve(self, context: Dict) -> Any:
+        """
+        Pull XCom value for the existing arg.
+        """
+        resolved_value = self._operator.xcom_pull(
+            context=context,
+            task_ids=[self._operator.task_id],
+            key=self._keys[0],
+            dag_id=self._operator.dag.dag_id,
+        )
+        if not resolved_value:
+            raise
+        return resolved_value[0]
+        # dict_result = {}
+        # i = dict_result
+        # for key in self._keys:
+        #     # Always set resolved value - will be overridden until last iteration
+        #     i[key] = resolved_value
+        #     i = i[key]
+        #
+        # return dict_result
+
+    # def __getitem__(self, key: str) -> 'XComArg':
+    #     """Return an XComArg for the specified key and the same operator as the current one.
+    # """
+    #     return XComArg(self._operator, [key])
+
+
+#  __iter__
+#   return x[1]
+#
+#
+#
+# xcom[restul]["a"]["b"]    xcom (keys= "result, a, ,b") -> 3
+# return {"a": {"b": 3}}
+#
+#
+# XcomCombine( a= Xcom, b=Xcom)
+
+
+def resolve_xcom_arg_class() -> Type[BaseXComArg]:
+    """Resolves custom XComArg class"""
+    path_to_custom_class = conf.get("core", "xcom_arg_class")
+    if path_to_custom_class:
+        clazz = import_string(path_to_custom_class)
+        base = import_string("airflow.models.xcom_arg.BaseXComArg")
+        if not issubclass(clazz, base):
+            raise TypeError(
+                f"Your custom XComArg class `{clazz.__name__}` is not a subclass of `{BaseXComArg.__name__}`."
+            )
+        return clazz
+
+    return BaseXComArg
+
+
+XComArg = resolve_xcom_arg_class()
```

Comment:
👍 

Not sure if XComArg is the best place to abstract those away btw. As I argued in https://github.com/apache/airflow/issues/8059 I do think its better to make XCom class itself be the abstract one instead of XComArg. If we abstract XComArg there may be inconsistencies on behaviour with existing XCom implementation. 

Making XCom a pluggable API would probs be better imo.

Reply:
+1 to Making Xcom backend pluggable with the default of airflow.models.xcom.XCom

Reply:
I agree

Reply:
In general, I think this can be done as separate PR to minimize scope. @casassg @evgenyshulman

Reply:
Moved this change to https://github.com/databand-ai/airflow/pull/4